File: python/0211.py
class WordDictionary:

    def __init__(self):
        self.trie = {}

    def addWord(self, word: str) -> None:
        t = self.trie
        for w in word:
            if w not in t:
                t[w] = {}
            t = t[w]
        t['#'] = True

    def search(self, word: str) -> bool:
        def check(idx, trie):
            if idx >= len(word):
                return '#' in trie
            c = word[idx]
            if c == '.':
                for (k, t) in trie.items():
                    if k == '#':
                        continue
                    if check(idx+1, t):
                        return True
            else:
                if c not in trie:
                    return False
                else:
                    if check(idx+1, trie[c]):
                        return True
            return False
        return check(0, self.trie)


obj = WordDictionary()
obj.addWord('bad')
obj.addWord('dad')
obj.addWord('mad')
print(obj.search('pad'))
print(obj.search('bad'))
print(obj.search('.ad'))
print(obj.search('b..'))

File: cpp/0211.cpp
#include <string>
#include <iostream>

using namespace std;

struct TrieNode {
    TrieNode* child[26];
    bool isWord;
};


class WordDictionary {
public:
    WordDictionary() {
        root = new TrieNode();
    }

    void addWord(string word) {
        TrieNode* node = root;
        for (const auto& c: word) {
            if (!node->child[c - 'a']) {
                node->child[c - 'a'] = new TrieNode();
            }
            node = node->child[c - 'a'];
        }
        node->isWord = true;
    }

    bool search(string word) {
        cout << word << ": ";
        return exactSearch(word, 0, root);
    }

    // bool search(string word) {
    //     cout << word << ": ";
    //     TrieNode* node = root;
    //     for (const auto& c: word) {
    //         node = node->child[c - 'a'];
    //         if (!node) return false;
    //     }
    //     return node->isWord;
    // }

    bool exactSearch(string& word, int idx, TrieNode* node) {
        if (!node) return false;
        for (int i = idx; i < word.size(); i++) {
            char c = word[i];
            if (c == '.') {
                // if (i == word.size() - 1) return true;
                for (int j = 0; j < 26; j++) {
                    if (exactSearch(word, i+1, node->child[j])) {
                        return true;
                    }
                }
                return false;
            } else {
                node = node->child[c - 'a'];
                if (!node) return false;
            }
        }
        return node->isWord;
    }
private:
    TrieNode* root;
};

int main() {
    WordDictionary wordDictionary = WordDictionary();

    // wordDictionary.addWord("bad");
    // wordDictionary.addWord("dad");
    // wordDictionary.addWord("mad");

    // cout << wordDictionary.search("pad") << endl; // return False
    // cout << wordDictionary.search("bad") << endl; // return True
    // cout << wordDictionary.search(".ad") << endl; // return True
    // cout << wordDictionary.search("b..") << endl; // return True

    wordDictionary.addWord("a");
    wordDictionary.addWord("ab");

    cout << wordDictionary.search("") << endl;
    cout << wordDictionary.search("a") << endl;
    cout << wordDictionary.search("ab") << endl;
    cout << wordDictionary.search("a") << endl;
    cout << wordDictionary.search("a.") << endl;
    cout << wordDictionary.search(".a") << endl;
    cout << wordDictionary.search(".b") << endl;
    cout << wordDictionary.search("ab.") << endl;
    cout << wordDictionary.search(".") << endl;
    cout << wordDictionary.search("..") << endl;

    return 0;
}
